### DISCUSSION TITLE
Question on building multiple Wasm files (and their js) for web

### DISCUSSION THREAD

**Discussion by ericoporto** [emscripten-core/emscripten#22483] [Q&A]
I would like to build an experiment in the browser using multiple wasm files.

In the ags engine we have a successful Emscripten build of the engine, I would like to also build the Script compiler, so I can make a webpage where you could edit AGS script files in the browser, run them in the compiler and then run in the engine.

Do I have to put both the script compiler and the engine in the same build or is there some way to smash together later/ load both wasm files? They also build a js files (and an html file), how would I proceed if it is possible to build them separately? 

If someone has any minimal example that plays with this idea, I would love to see it.

I am using the CMake build system to build each.

Thanks

**Comment by sbc100** [comment]
If you want to hook up two completely seperate programs then emscripten doesn't really have any features for doing this, would need to do all the hooking up yourself.  The one thing you would need to do is to compiler each program with `-sMODULARIZE` so that they didn't clobber each other in the global namespace.

If the engine and the compiler can be considered the same program then you could consider using dyanamic linking.  In this case you would choose one of the programs to be the main module (`-sMAIN_MODULE`) and the other to be a side module (`-sSIDE_MODULE`).   Using `-sMAIN_MODULE=2` is highly recommended for code size savings).

**Reply by ericoporto** [reply]
Uhm, stupid question, if I don't have any side modules and don't intend to dynamically load things, should I add `-sMAIN_MODULE` still and do I gain anything by adding `-sMAIN_MODULE=2` in my compilation? I don't have this anywhere in my build flags ... I can't tell from [the docs](https://emscripten.org/docs/tools_reference/settings_reference.html#settings-reference). [Here are my flags](https://github.com/adventuregamestudio/ags/blob/6ce71b446cb7a561dc4fbcec57f71a29547c5b1e/CMakeLists.txt#L145).

**Reply by sbc100** [reply]
If you don't want to link mutiple wasm modules (i.e. dynamic linking) then no you don't want to use `-sMAIN_MODULE`.  

I though this question was about linking multiple modules but perhaps I misunderstood.

**Reply by sbc100** [reply]
If your wasm modules are completely separate then instead of using dynamic linking you would need to hook them up yourself using their external APIs (this may include copying any data from the memory of one running program into the memory of the other).

**Reply by ericoporto** [reply]
ok, I think the way to go is to build a third program that uses both the compiler and engine as a library and statically link everything. Then I solve the interface with everything in this third program. At least, it sounds saner and I can keep the problem solving in c++ and not think too much about the details of the browser stuff.
